ledger/src/ui/src/pin.c

```diff
@@ -33,29 +35,66 @@
 #define IS_NUM(c) IS_IN_RANGE(c, '0', '9')
 #define IS_ALPHANUM(c) (IS_ALPHA(c) || IS_NUM(c))
 
-/*
- * Validates that the pin has exactly PIN_LENGTH alphanumeric characters
- * with at least one alphabetic character.
- *
- * @arg[in] pin null-terminated string representing the pin to validate
- * @ret     true if pin is valid, false otherwise
- */
-bool is_pin_valid(unsigned char *pin) {
+bool is_pin_valid(pin_t* pin_ctx) {
     // PIN_LENGTH is the only length accepted
-    size_t length = strnlen((const char *)pin, PIN_LENGTH + 1);
-    if (length != PIN_LENGTH) {
+    size_t length =
+        strnlen((const char*)pin_ctx->pin_buffer.payload, MAX_PIN_LENGTH + 1);
+    if (length != MAX_PIN_LENGTH) {
         return false;
     }
     // Check if PIN is alphanumeric
     bool hasAlpha = false;
-    for (int i = 0; i < PIN_LENGTH; i++) {
-        if (!IS_ALPHANUM(pin[i])) {
+    for (int i = 0; i < MAX_PIN_LENGTH; i++) {
+        if (!IS_ALPHANUM(pin_ctx->pin_buffer.payload[i])) {
             return false;
         }
-        if (hasAlpha || IS_ALPHA(pin[i])) {
+        if (hasAlpha || IS_ALPHA(pin_ctx->pin_buffer.payload[i])) {
             hasAlpha = true;
         }
     }
 
     return hasAlpha;
+}
+
+void get_pin_ctx(pin_t* pin_ctx, unsigned char* pin_buffer) {
+    memcpy(pin_buffer, pin_ctx->pin_raw, sizeof(pin_ctx->pin_raw));
+}
+
+void set_pin_ctx(pin_t* pin_ctx, unsigned char* pin_buffer) {
+    memcpy(pin_ctx->pin_raw, pin_buffer, sizeof(pin_ctx->pin_raw));
+}
```

Comment:
Yeah, there isn't an easy solution to this. We need a copy of the pin outside `G_bolos_ux_context` since the context is reset on every new command. Your approach works, I'll change to that.